mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp

```diff
@@ -180,42 +180,79 @@ struct MFMAOpLowering : ConvertToLLVMPattern {
                        immValues[0], immValues[1], immValues[2]});
     } else if (mfmaInstr.endswith("bf16")) {
       // BF16.
-      Type castedVectorType = VectorType::get({2}, rewriter.getIntegerType(16));
-      Type castedLLVMVectorType = typeConverter->convertType(castedVectorType);
-      Value castedSourceA = rewriter.create<LLVM::BitcastOp>(
-          op->getLoc(), castedLLVMVectorType, adaptor.sourceA());
-      Value castedSourceB = rewriter.create<LLVM::BitcastOp>(
-          op->getLoc(), castedLLVMVectorType, adaptor.sourceB());
       if (mfmaInstr == "mfma_f32_32x32x2bf16")
         rewriter.replaceOpWithNewOp<ROCDL::mfma_f32_32x32x2bf16>(
             op, adaptor.destC().getType(),
-            ValueRange{castedSourceA, castedSourceB, adaptor.destC(),
+            ValueRange{adaptor.sourceA(), adaptor.sourceB(), adaptor.destC(),
                        immValues[0], immValues[1], immValues[2]});
       else if (mfmaInstr == "mfma_f32_32x32x4bf16")
         rewriter.replaceOpWithNewOp<ROCDL::mfma_f32_32x32x4bf16>(
             op, adaptor.destC().getType(),
-            ValueRange{castedSourceA, castedSourceB, adaptor.destC(),
+            ValueRange{adaptor.sourceA(), adaptor.sourceB(), adaptor.destC(),
                        immValues[0], immValues[1], immValues[2]});
       else if (mfmaInstr == "mfma_f32_16x16x8bf16")
         rewriter.replaceOpWithNewOp<ROCDL::mfma_f32_16x16x8bf16>(
             op, adaptor.destC().getType(),
-            ValueRange{castedSourceA, castedSourceB, adaptor.destC(),
+            ValueRange{adaptor.sourceA(), adaptor.sourceB(), adaptor.destC(),
                        immValues[0], immValues[1], immValues[2]});
       else if (mfmaInstr == "mfma_f32_16x16x2bf16")
         rewriter.replaceOpWithNewOp<ROCDL::mfma_f32_16x16x2bf16>(
             op, adaptor.destC().getType(),
-            ValueRange{castedSourceA, castedSourceB, adaptor.destC(),
+            ValueRange{adaptor.sourceA(), adaptor.sourceB(), adaptor.destC(),
                        immValues[0], immValues[1], immValues[2]});
       else if (mfmaInstr == "mfma_f32_4x4x2bf16")
         rewriter.replaceOpWithNewOp<ROCDL::mfma_f32_4x4x2bf16>(
             op, adaptor.destC().getType(),
-            ValueRange{castedSourceA, castedSourceB, adaptor.destC(),
+            ValueRange{adaptor.sourceA(), adaptor.sourceB(), adaptor.destC(),
                        immValues[0], immValues[1], immValues[2]});
     }
 
     return success();
   }
 };
+
+struct BFOpLowering : ConvertToLLVMPattern {
+  explicit BFOpLowering(MLIRContext *context, LLVMTypeConverter &typeConverter)
+      : ConvertToLLVMPattern(gpu::BFConvertOp::getOperationName(), context,
+                             typeConverter) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto bfOp = cast<gpu::BFConvertOp>(op);
```

Comment:
Would be good to have a comment with the math.

Reply:
added